content/Watsonx/WatsonxAI/files/rag-with-discovery.ipynb

```diff
@@ -0,0 +1,295 @@
+    "gen_params = {\n",
+    "    \"DECODING_METHOD\" : \"greedy\",\n",
+    "    \"MAX_NEW_TOKENS\" : 500,\n",
+    "    \"MIN_NEW_TOKENS\" : 20,\n",
+    "    \"STREAM\" : False,\n",
+    "    \"TEMPERATURE\" : 1.7,\n",
+    "    \"TOP_K\" : 50,\n",
+    "    \"TOP_P\" : 1,\n",
+    "    \"RANDOM_SEED\" : 10\n",
+    "}\n",
+    "\n",
+    "model = Model( model_id, credentials, gen_params, watsonx_project_id )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 51,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "Answer:\n",
+      "€2,250 million.'] ### As of March 31, 2023, the Company had €2,250 million of outstanding long- term debt and approximately €42 million of accrued interest designated as non-derivative hedges of its net investment in certain foreign operations.\n"
+     ]
+    }
+   ],
+   "source": [
+    "#######################################################################################\n",
+    "prompt_template = \"\"\"\n",
+    "Article:\n",
+    "###\n",
+    "%s\n",
+    "###\n",
+    "\n",
+    "Answer the following question using only information from the article. \n",
+    "Answer in a complete sentence, with proper capitalization and punctuation. \n",
+    "If there is no good answer in the article, say \"I don't know\".\n",
+    "\n",
+    "Question: %s\n",
+    "Answer: \n",
+    "\"\"\"\n",
+    "\n",
+    "#######################################################################################\n",
+    "def augment( template_in, context_in, query_in ):\n",
+    "    return template_in % ( context_in,  query_in )\n",
+    "\n",
+    "\n",
+    "#######################################################################################\n",
+    "import json\n",
+    "\n",
+    "def generate( model_in, augmented_prompt_in ):\n",
+    "    \n",
+    "    generated_response = model_in.generate( augmented_prompt_in )\n",
+    "\n",
+    "    if ( \"results\" in generated_response ) \\\n",
+    "       and ( len( generated_response[\"results\"] ) > 0 ) \\\n",
+    "       and ( \"generated_text\" in generated_response[\"results\"][0] ):\n",
+    "        return generated_response[\"results\"][0][\"generated_text\"]\n",
+    "    else:\n",
+    "        print( \"The model failed to generate an answer\" )\n",
+    "        print( \"\\nDebug info:\\n\" + json.dumps( generated_response, indent=3 ) )\n",
+    "        return \"\"\n",
+    "\n",
+    "\n",
+    "########################################################################################\n",
+    "import re\n",
+    "\n",
+    "augmented_prompt = augment( prompt_template, combined_disc_results, discovery_question )\n",
+    "output = generate( model, augmented_prompt )\n",
+    "if not re.match( r\"\\S+\", output ):\n",
+    "    print( \"The model failed to generate an answer\")\n",
+    "print( \"\\nAnswer:\\n\" + output )\n"
+   ]
```

Comment:
I am not sure how to remove the `€2,250 million.. ` at the beginning. Maybe it will be fixed with a repetition penalty.
<img width="283" alt="Screenshot 2023-10-03 at 12 35 35 PM" src="https://github.com/ibm-build-lab/VAD-VAR-Workshop/assets/19367627/1c1924d4-3b83-4e90-84e0-86c978f61720">

Reply:
Like you mentioned this can be addressed by tuning the `gen_parameters` that you provide to the watsonx.ai model, I went ahead and updated the `MIN_TOKEN` param to reduce the amount output for the LLM, making it more concise. `€2,250 million` is the expected answer.